### ISSUE TITLE
Transaction search not returning embedded transactions

### ISSUE THREAD

**Issue by rg911** [nemtech/catapult-rest#351]
Transaction search not return embedded transactions even embedded flag = true

**Comment by Vektrat** [comment]
i just tried it and it's working

**Comment by Vektrat** [comment]
btw if you expect the endpoint to return the embedded transactions nested within the object, this is not how the endpoint works, for this you need to call the detail endpoint on a single transaction

**Comment by rg911** [comment]
Doesn't work for me

http://localhost:3000/transactions?address=TAE6WU3YJMFIYNKCMENY5ZYA4BTO7W5A5AE5Q452&embedded=true

Payload:
```
{
    "data": [
        {
            "meta": {
                "height": "1",
                "hash": "9012FEEDEB25DD8C2B7A9EA5DFC35565388106AF0CCF7E2F0E1A4E5B94FF5FA9",
                "merkleComponentHash": "9012FEEDEB25DD8C2B7A9EA5DFC35565388106AF0CCF7E2F0E1A4E5B94FF5FA9",
                "index": 12
            },
            "transaction": {
                "signature": "386E3E04489A9153F42FB3942914007EF15C266D4BCE173567AF4B0A4DC447AB6F06CF5226F293039D8A04D3C1583EB66C5171B279D910F5FF35009D8B7A9709",
                "signerPublicKey": "C0225DCE7D5AB2BE2A1548B0EF49CD82B3CE9EA68E52AFDABB49AED3525A762D",
                "version": 1,
                "network": 152,
                "type": 16724,
                "maxFee": "0",
                "deadline": "1",
                "recipientAddress": "9809EB53784B0A8C3542611B8EE700E066EFDBA0E809D873BA",
                "mosaics": [
                    {
                        "id": "85BBEA6CC462B244",
                        "amount": "449949999900000"
                    },
                    {
                        "id": "941299B2B7E1291C",
                        "amount": "3750000"
                    }
                ]
            },
            "id": "5ECFDCDEC08B395CD259E015"
        },
        {
            "meta": {
                "height": "209",
                "hash": "ABC86310606E667AFF8BF0A146A0FAF22CE29BE1BB34DE920F99B154A1F916D4",
                "merkleComponentHash": "ABC86310606E667AFF8BF0A146A0FAF22CE29BE1BB34DE920F99B154A1F916D4",
                "index": 0
            },
            "transaction": {
                "signature": "59758C51FF1A62D5A986CD8DF4535C830D286141A817CB4F2595DE3EF7D425E83F86FB8E5DED61658C2E34807E6A6DC25093E2ACDD8F6CA84CAAF9B880660703",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16717,
                "maxFee": "1000000",
                "deadline": "17334501269",
                "nonce": 159833969,
                "id": "22AAE7DB4C6B393B",
                "flags": 7,
                "divisibility": 3,
                "duration": "1000"
            },
            "id": "5ED1094DB455912BAA4B1139"
        },
        {
            "meta": {
                "height": "210",
                "hash": "2F311CD66B0F65C20C11D95B60BFBD011F56AA4A12DE4FCA27CC2745240A6BD9",
                "merkleComponentHash": "2F311CD66B0F65C20C11D95B60BFBD011F56AA4A12DE4FCA27CC2745240A6BD9",
                "index": 0
            },
            "transaction": {
                "signature": "94CD8E068B0F1F6500A9AF48F41E7C3BA0BD8649D514857EB6DEFD81CC8676A7D4DEFA9FB389E5632EC389D9B274CBADA3AE7A549AF79EAC24EA5FC31A925900",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16705,
                "maxFee": "1000000",
                "deadline": "17334509768",
                "transactionsHash": "B51B5AB8E6DA61267646D0A63AE8765C5F0EB250F2E5BE1F49C3598234B93E9D",
                "cosignatures": []
            },
            "id": "5ED10953B455912BAA4B113F"
        },
        {
            "meta": {
                "height": "211",
                "hash": "61B220A8B46BD3FCFD1F141998A13F6D4DEABC062A90ADA4B6468C22754A64AE",
                "merkleComponentHash": "61B220A8B46BD3FCFD1F141998A13F6D4DEABC062A90ADA4B6468C22754A64AE",
                "index": 0
            },
            "transaction": {
                "signature": "BD549D792F589EE47CCE140DA790ED654C2DBF80336FBC7F8AE7F66ACB1F437BD8DE8E85C2F1CDFFFD5BD303161230F2DF865B26F3CD47D3DC73FCA1DBF89A09",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16705,
                "maxFee": "1000000",
                "deadline": "17334515906",
                "transactionsHash": "9B8982E2C6A196DBCC031EC1110EE8D3A3097C5178CC548B8C7BB20F8D67FFFD",
                "cosignatures": []
            },
            "id": "5ED10960B455912BAA4B1146"
        },
        {
            "meta": {
                "height": "212",
                "hash": "AD702E0C4B2F8A431880B310EC5C2F4C477BF1FE27CED69201BFEB01305A8019",
                "merkleComponentHash": "AD702E0C4B2F8A431880B310EC5C2F4C477BF1FE27CED69201BFEB01305A8019",
                "index": 0
            },
            "transaction": {
                "signature": "C94DAFE1657924076C9336DC79ED21EF62DF344667E908C8B4CF463587F3695A0FFE27A9419A03CF07CE7D41E4F660F4692A563293A76286D460920941056C04",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16705,
                "maxFee": "1000000",
                "deadline": "17334528144",
                "transactionsHash": "98D5EB5DB29111F209407260E544B9ED6DE7D3CA090A38245919F9FFDD59F031",
                "cosignatures": []
            },
            "id": "5ED1096CB455912BAA4B114C"
        },
        {
            "meta": {
                "height": "214",
                "hash": "23D74F2BE6541B3482C6DA613028D7FCFF2DAB9CD64960891EF18A43F289F6BA",
                "merkleComponentHash": "23D74F2BE6541B3482C6DA613028D7FCFF2DAB9CD64960891EF18A43F289F6BA",
                "index": 0
            },
            "transaction": {
                "signature": "D8DAD92283534ACE2909F100F8AFAC56A44599ABB49EB209EC3BDC6155726D1C8978AAA1CFD924EF3C37A092F4B2EE651587FB3F7212CC7366E3709813CAFA0A",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16718,
                "maxFee": "1000000",
                "deadline": "17334540448",
                "registrationType": 0,
                "duration": "10",
                "id": "B55C04BB034B085C",
                "name": "root-test-namespace-3021"
            },
            "id": "5ED10978B455912BAA4B1154"
        },
        {
            "meta": {
                "height": "215",
                "hash": "E8F7F50FB95EFCE87CB2216E086D38D2F9F739C731A27C57F8EA3C25159BB787",
                "merkleComponentHash": "E8F7F50FB95EFCE87CB2216E086D38D2F9F739C731A27C57F8EA3C25159BB787",
                "index": 0
            },
            "transaction": {
                "signature": "7834F0403DB04B22D0BCE0E3B8F6233953F78C4123C702D25901FBF80B767EC4F3B7C2EB06F44979352ADF4AD15E3788EF76C286971C71050ED59B113543CA0D",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16705,
                "maxFee": "1000000",
                "deadline": "17334552688",
                "transactionsHash": "055C7133CCE272CDD2E303782FE1A46FA96C303CD610A46F5FC76C7D7F5C112A",
                "cosignatures": []
            },
            "id": "5ED1097EB455912BAA4B115B"
        },
        {
            "meta": {
                "height": "216",
                "hash": "4D2514E0AE240C5ED49A8CDF6315C7CFC2B8CF9307EE596F7A6E6FCF0790A8CE",
                "merkleComponentHash": "4D2514E0AE240C5ED49A8CDF6315C7CFC2B8CF9307EE596F7A6E6FCF0790A8CE",
                "index": 0
            },
            "transaction": {
                "signature": "6F8AD344AB311990D8B93799F26D1B19EBF809FA430108C7A5358DA47B9FE7055B8FD19776BD0A438F65D9D568B4FC7F7A8722B64E1FB0DD738B6CDD6E33CF02",
                "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                "version": 1,
                "network": 152,
                "type": 16705,
                "maxFee": "1000000",
                "deadline": "17334558820",
                "transactionsHash": "316986BFB9EF9A8312566FD8B8C6F4593E285B181F972BCCE76AE0855EB416F3",
                "cosignatures": []
            },
            "id": "5ED1098AB455912BAA4B1162"
        }
    ],
    "pagination": {
        "totalEntries": 8,
        "pageNumber": 1,
        "pageSize": 20,
        "totalPages": 1
    }
}
```

**Comment by rg911** [comment]
> btw if you expect the endpoint to return the embedded transactions nested within the object, this is not how the endpoint works, for this you need to call the detail endpoint on a single transaction

Don't think that's right unless the OpenAPI spec is wrong with the embedded flag: `When true, the endpoint also returns all the embedded aggregate transactions. Otherwise, only top-level transactions used to calcualate the block transactionsHash are returned.`

Otherwise this flag does not do anything.

**Comment by Vektrat** [comment]
both this and openapi are right

all the transactions are in the same collection, embedded and not embedded, by default the endpoint ignores embedded transactions, but if you enable it, you will get a result of mixed parent and embedded transactions

**Comment by rg911** [comment]
> both this and openapi are right
> 
> all the transactions are in the same collection, embedded and not embedded, by default the endpoint ignores embedded transactions, but if you enable it, you will get a result of mixed parent and embedded transactions

see the example payload, they are not in there

**Comment by Vektrat** [comment]
i'm not sure we're understanding each other

you will never get a `transactions` property with an array of embedded transactions in this endpoint, i can't know if this endpoint is working or not from payload because i don't know what other data is in the database so i don't know if embedded exist for this address

**Comment by rg911** [comment]
pick one of the aggregate for example : 

http://localhost:3000/transactions/5ED1098AB455912BAA4B1162

```
{
    "meta": {
        "height": "216",
        "hash": "4D2514E0AE240C5ED49A8CDF6315C7CFC2B8CF9307EE596F7A6E6FCF0790A8CE",
        "merkleComponentHash": "4D2514E0AE240C5ED49A8CDF6315C7CFC2B8CF9307EE596F7A6E6FCF0790A8CE",
        "index": 0,
        "id": "5ED1098AB455912BAA4B1162"
    },
    "transaction": {
        "signature": "6F8AD344AB311990D8B93799F26D1B19EBF809FA430108C7A5358DA47B9FE7055B8FD19776BD0A438F65D9D568B4FC7F7A8722B64E1FB0DD738B6CDD6E33CF02",
        "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
        "version": 1,
        "network": 152,
        "type": 16705,
        "maxFee": "1000000",
        "deadline": "17334558820",
        "transactionsHash": "316986BFB9EF9A8312566FD8B8C6F4593E285B181F972BCCE76AE0855EB416F3",
        "cosignatures": [],
        "transactions": [
            {
                "meta": {
                    "height": "216",
                    "aggregateHash": "4D2514E0AE240C5ED49A8CDF6315C7CFC2B8CF9307EE596F7A6E6FCF0790A8CE",
                    "aggregateId": "5ED1098AB455912BAA4B1162",
                    "index": 0,
                    "id": "5ED1098AB455912BAA4B1163"
                },
                "transaction": {
                    "signerPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                    "version": 1,
                    "network": 152,
                    "type": 17220,
                    "targetPublicKey": "57B76C16D6CF24E2605D6F08D6C6B84B9D5378A880D98C8C14108EFABBED6551",
                    "scopedMetadataKey": "0000000000000005",
                    "targetNamespaceId": "B55C04BB034B085C",
                    "valueSizeDelta": 10,
                    "valueSize": 10,
                    "value": "00000000000000000000"
                }
            }
        ]
    }
}
```

I don't see the embedded in the search payload above. 
In fact the setting embedded to true or false makes no difference. 

What's the transaction search payload looks like on your end? Can you post here?

**Comment by martinayora** [comment]
I have tried just now. This is an example with embedded false and embedded true. Height and signerPublicKey are used to narrow the number of transactions only.

In the first case only two transactions are retrieved. They are both aggregate complete transactions. As I'm not requesting embedded ones, only aggregate are retrieved.

In the second case embedded is true, so both aggregate and the inner transactions which are namespace metadata transactions, are retrieved.

```
Query:
127.0.0.1:3000/transactions?height=79&signerPublicKey=80f8f00dda6c6d3a46b70bee66752bf49f6f97880b7d3917454287b892033a2e&embedded=false

Output:
{
    "data": [
        {
            "meta": {
                "height": "79",
                "hash": "5D1823A4347F9C3D42B3D32231288DAAAD4D83D879C4FCD4DD070976C39D915A",
                "merkleComponentHash": "5D1823A4347F9C3D42B3D32231288DAAAD4D83D879C4FCD4DD070976C39D915A",
                "index": 0
            },
            "transaction": {
                "signature": "6DAB3845BE2967EEC5E2921BD6102EB70BC8BCDF2A6C21D78507857A358F70EE91EE81CAA36996412D0239680948780576452211D574E65987586A738C24120A",
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 16705,
                "maxFee": "26664",
                "deadline": "17088664880",
                "transactionsHash": "27CBCA5D29365C538E51C9720C1EB1042A11DD50FA907C51E2C06787F6E48535",
                "cosignatures": []
            },
            "id": "5ECD571E8D1ABC1639297EED"
        },
        {
            "meta": {
                "height": "79",
                "hash": "D202057DDEC3F4A9AAB5C1F5654D1755AF5B4EDE1DF5437B1B6325EA0A16C422",
                "merkleComponentHash": "D202057DDEC3F4A9AAB5C1F5654D1755AF5B4EDE1DF5437B1B6325EA0A16C422",
                "index": 3
            },
            "transaction": {
                "signature": "F0F7B92BAA3A14EF21A3C137D73B5A368CF63E00A6306DCBAB3FAD62F1F26326129E6CBE6C552DF28B6A74C58453B0FC3AD78C36944999123A3E1CA2B9E25302",
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 16705,
                "maxFee": "26664",
                "deadline": "17088664881",
                "transactionsHash": "59F53E5936300CC591A6BD9FF4F4079BF271D66EC5F5B052A8FCFB8D86D98016",
                "cosignatures": []
            },
            "id": "5ECD571E8D1ABC1639297EF3"
        }
    ],
    "pagination": {
        "totalEntries": 2,
        "pageNumber": 1,
        "pageSize": 20,
        "totalPages": 1
    }
}



Query:
127.0.0.1:3000/transactions?height=79&signerPublicKey=80f8f00dda6c6d3a46b70bee66752bf49f6f97880b7d3917454287b892033a2e&embedded=true

Output:
{
    "data": [
        {
            "meta": {
                "height": "79",
                "hash": "5D1823A4347F9C3D42B3D32231288DAAAD4D83D879C4FCD4DD070976C39D915A",
                "merkleComponentHash": "5D1823A4347F9C3D42B3D32231288DAAAD4D83D879C4FCD4DD070976C39D915A",
                "index": 0
            },
            "transaction": {
                "signature": "6DAB3845BE2967EEC5E2921BD6102EB70BC8BCDF2A6C21D78507857A358F70EE91EE81CAA36996412D0239680948780576452211D574E65987586A738C24120A",
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 16705,
                "maxFee": "26664",
                "deadline": "17088664880",
                "transactionsHash": "27CBCA5D29365C538E51C9720C1EB1042A11DD50FA907C51E2C06787F6E48535",
                "cosignatures": []
            },
            "id": "5ECD571E8D1ABC1639297EED"
        },
        {
            "meta": {
                "height": "79",
                "aggregateHash": "5D1823A4347F9C3D42B3D32231288DAAAD4D83D879C4FCD4DD070976C39D915A",
                "aggregateId": "5ECD571E8D1ABC1639297EED",
                "index": 0
            },
            "transaction": {
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 17220,
                "targetAddress": "9097FBB41E6FD46C1AF880FCA4503E68E7510B2A6AA8B510",
                "scopedMetadataKey": "0000000000000001",
                "targetNamespaceId": "A6286EB7E7C454E9",
                "valueSizeDelta": 4,
                "valueSize": 4,
                "value": "04040404"
            },
            "id": "5ECD571E8D1ABC1639297EEE"
        },
        {
            "meta": {
                "height": "79",
                "hash": "D202057DDEC3F4A9AAB5C1F5654D1755AF5B4EDE1DF5437B1B6325EA0A16C422",
                "merkleComponentHash": "D202057DDEC3F4A9AAB5C1F5654D1755AF5B4EDE1DF5437B1B6325EA0A16C422",
                "index": 3
            },
            "transaction": {
                "signature": "F0F7B92BAA3A14EF21A3C137D73B5A368CF63E00A6306DCBAB3FAD62F1F26326129E6CBE6C552DF28B6A74C58453B0FC3AD78C36944999123A3E1CA2B9E25302",
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 16705,
                "maxFee": "26664",
                "deadline": "17088664881",
                "transactionsHash": "59F53E5936300CC591A6BD9FF4F4079BF271D66EC5F5B052A8FCFB8D86D98016",
                "cosignatures": []
            },
            "id": "5ECD571E8D1ABC1639297EF3"
        },
        {
            "meta": {
                "height": "79",
                "aggregateHash": "D202057DDEC3F4A9AAB5C1F5654D1755AF5B4EDE1DF5437B1B6325EA0A16C422",
                "aggregateId": "5ECD571E8D1ABC1639297EF3",
                "index": 0
            },
            "transaction": {
                "signerPublicKey": "80F8F00DDA6C6D3A46B70BEE66752BF49F6F97880B7D3917454287B892033A2E",
                "version": 1,
                "network": 144,
                "type": 17220,
                "targetAddress": "9097FBB41E6FD46C1AF880FCA4503E68E7510B2A6AA8B510",
                "scopedMetadataKey": "0000000000000001",
                "targetNamespaceId": "A6286EB7E7C454E9",
                "valueSizeDelta": -3,
                "valueSize": 4,
                "value": "05040404"
            },
            "id": "5ECD571E8D1ABC1639297EF4"
        }
    ],
    "pagination": {
        "totalEntries": 4,
        "pageNumber": 1,
        "pageSize": 20,
        "totalPages": 1
    }
}
```

**Comment by fboucquez** [comment]
Would it be possible to update the behavior of "embedded" to put embedded transaction inside aggregate or maybe change it to an enum like "top, embbeded_flat, embbeded_inner" giving the client the option of including the inner transaction inside the aggregate? This "flat" embedded change will make life harder for users requiring more rest calls to recreate the original aggregate transactions.

@dgarcia360

**Comment by Vektrat** [comment]
we won't be making this change, this has been extensively discussed with @dgarcia360 and we agreed that it's bad API design and may rise problems, sadly i think the thread is lost due to the slack's limit (i regret not having done so in github and we should always do

if you are curious on the details i can list them here again as they come to me, but don't expect any changes to this endpoint in this regard

**Comment by fboucquez** [comment]
Can you explain why it is a bad design? Are you sure is it not an implementation constraint? 

And why is NO always the default answer?

**Comment by Vektrat** [comment]
- it was done like this before, so it's not an implementation constraint
- we have to manually build the response: first by doing a query to the database to get the parent transactions, then do yet another query to get the nested transactions, and then manually go over them filling the sub-array (this process can be altered but it's still multiple queries per transaction to the database, or iterating over results to build it
- the `/transactions` endpoint is a LIST endpoint, the goal of this is to list and navigate over large lists of elements, if you need information from a specific element, you can/should call the DETAIL endpoint (which is `/transactions/:id`
- we are not dropping functionality by removing this
- we believe most of the times getting ALL the embedded for ALL the nested in the collection won't be necessary, and the drawback is multiple queries to the database and manual build of the result
- this result escapes one of the goals of pagination: with pagination you can have control over the result's size, but if the results are so variable, it would be possible that we were returning responses with hundreds of transactions rather than fixed 20 to say something

all in all, functionality is the same, performance is greatly considered, and this model is closer to what a standard REST API with list/detail endpoints should be - we cannot see any real advantage to how this endpoint was handled before, and instead we see a lot of advantages from the suggested solution

NO is never the default answer, but unlike you we just don't give YES by default and at any cost without first deeply considering our decisions

**Comment by fboucquez** [comment]
When did I say yes without delivering?

**Comment by Vektrat** [comment]
can we close this issue then?

**Comment by rg911** [comment]
> can we close this issue then?

Leave it open until next week. I don't get the similar payload as @martinayora posted on my end by calling the endpoint directly

**Comment by fboucquez** [comment]
Clients and SDKs would need to adapt performing all the extra calls

**Comment by Vektrat** [comment]
> Leave it open until next week. I don't get the similar payload as @martinayora posted on my end by calling the endpoint directly

what is different?

> Clients and SDKs would need to adapt performing all the extra calls

and this was expected, the updates are not a "pagination" update as it is being called, it's a review of the API, and thus all the endpoints are being analyzed and changed, pagination is just one of the things that changed, and it is being shipped together because splitting it made no sense and would have implied extra work - i don't see why SDKs need to perform extra calls (all? just one, if they weren't doing it already), i think this is on the clients side

**Comment by rg911** [comment]
- Separate the embedded & aggregate will cause confusion if the embedded flag set to the default (not provided). Multiple API calls to fulfill a single aggregate is not user friendly though it benefits the performance on rest.

- @martinayora 's example works only because the aggregate signer is the same as the inner transaction signer. If the inner tx has got different signer (publicKey) it will not return. This means even if the `embedded` flag is set to `true` the rest may not even return the full aggrege transaction set ( (e.g. my case above)). It also depends on the other search criteria used int he query.  Not only even causing more confusion here but seems a bug to me??.

**Comment by martinayora** [comment]
If the inner has different signerPublicKey and I'm filtering by signerPublicKey, they won't be returned that is correct. This endpoints shows transactions filtered by the given params. I used this filter and the height because it narrowed the response. But it works exactly the same without the filter, it only returns 154 transactions instead of two. And 169 when using embedded = true instead of 4.

When using the address filter it is looking into `meta.addresses` field in mongo. Is it possible that the address you are using is in the parent transaction but not in the inner ones? Maybe this is why your response doesn't change when using embedded. Can you run it with out any filter (just embedded true and false) and see if the pagination total entries value changes?

**Comment by rg911** [comment]
> If the inner has different signerPublicKey and I'm filtering by signerPublicKey, they won't be returned that is correct. This endpoints shows transactions filtered by the given params. I used this filter and the height because it narrowed the response. But it works exactly the same without the filter, it only returns 154 transactions instead of two. And 169 when using embedded = true instead of 4.
> 
> When using the address filter it is looking into `meta.addresses` field in mongo. Is it possible that the address you are using is in the parent transaction but not in the inner ones? Maybe this is why your response doesn't change when using embedded. Can you run it with out any filter (just embedded true and false) and see if the pagination total entries value changes?

@martinayora I think that's the cause of the issue. My point here is the aggregate transaction should be treated specifically as the search criteria should not be applied on inner transactions if it has met on aggregate wrapper already.

**Comment by dgarcia360** [comment]
Links to decisions taken:
- https://github.com/nemtech/symbol-openapi/pull/92#issuecomment-615216509
-https://github.com/nemtech/symbol-openapi/pull/92#issuecomment-615357643
- https://github.com/nemtech/symbol-openapi/pull/92#issuecomment-617701965

> @fboucquez Can you explain why it is a bad design?
>  @rg911 This can be confusing to the end user and not very user friendly.
> @fboucquez Are you sure is it not an implementation constraint?

The main reason for this change is to enhance the developer experience. If a user queries for transactions with a given filter (e.g., an exchange retrieving all deposits received to their account by type), they would have to unwrap all the aggregate transactions and only pick the ones that match the desired filter. This means, in practice, that the filter applied on REST was practically useless since the dev had to unwrap any aggregate transaction returned and apply the filter again manually in their application.

Secondly, REST was adding the embedded transactions inside every aggregate for every call,  but MongoDB is storing them individually. When searching with filters, returning aggregate queries with embedded transactions was much less performant, since REST had to query the individual transaction first, and then construct the aggregate with all the unrelated embedded txs inside.

All in all, I believe the previous approach was the implementation constraint because it was making too complex filtering the transactions endpoint. I understand the previous way can be less confusing because it's the way of doing we are used to, but not the best option. Let me know if you feel the same after reading the message.

> @fboucquez  Clients and SDKs would need to adapt to performing all the extra calls
> @rg911 It means after the search,  potentially multiple extra API calls are needed to get details of the embedded.

When you ask for a list of blocks, you only expect to get the block header. If you want to know about the contents of this block, you query the individual details associated. The same could apply to aggregate transactions; the details of each aggregate transaction could only be requested when required.

If the query is done with the ``embedded`` flag set to false, the SDK would only have to return the Aggregate Transaction Header. If the ``embedded`` flag is set to true, the SDK should return each inner transaction separately, as it comes from the API.

You can check as an example of this behavior with the explorer. They don't require the inner transactions at all to list transactions, only when a certain transaction is clicked: http://explorer-951.symboldev.network/transactions. 

Still, if the endpoint needs to be used for data processing (e.g. analytics), it feels more useful to return each transaction individually rather than having to unwrap each aggregate to inspect its content.

> @fboucquez Would it be possible to update the behavior of "embedded" to put embedded transaction inside aggregate or maybe change it to an enum like "top, embbeded_flat, embbeded_inner" giving the client the option of including the inner transaction inside the aggregate? 

To me, filters + aggregates are just not compatible since REST would return aggregate with unrelated transactions. For this reason, I would not offer an ``embedded_inner`` option.

**Comment by dgarcia360** [comment]
> Search criteria has been applied on both Aggregate Transaction and embedded transactions. This is causing even more confusion. E.g. when Address is used as a search criteria together with including embedded enabled (embedded flag = true), the result might return only partial embedded transaction list since that address might not match the address criteria on inner transactions. IMO, this seems a bug to me.

This looks fine and desirable to me. Again, imagine an exchange filtering transfer transactions they receive. They would only be interested in the inner transfer transactions, not the aggregate at all. If they want to review the parent aggregate, they know this is an embedded transfer transaction, the hash assigned to the parent aggregate, and index within the aggregate.

**Comment by Vektrat** [comment]
thanks for the extremely detailed explanation, and for finding the references to discussion, they were in github after all ^^

**Comment by fboucquez** [comment]
Having an incomplete aggregate transaction is problematic:
1) It cannot be catbuffer serialized
2) The calculated size would be incorrect
3) The effective_fee and recommended fees cannot be calculated as the size is incorrect 
4) It cannot be resigned or rehashed

If you load a page with embedded=true, it's not guaranteed that the inner transaction are in the same page. Re-loading each aggregate transaction from a page would be a must for those cases too (there may be others I haven't got time to fully test the consequences).

I agree that Res needs to control the amount of data return and users may want to filter inner transactions. I'm just asking if having the option embedded_inner  when opted-in by the user is possible like it used to. I would like to treat the aggregate transaction like an immutable whole object (this not a bad design).  As you stated, having the transaction split is also due to how they are stored in the database. I wouldn't filter out inner transactions if the parent aggregate transaction matches the query param criteria. This option doesn't exclude any of the other ones

**Comment by rg911** [comment]
Also it would be better to add the `numEmbeddedTransactions` or other forms of indicator in the aggregate transaction meta to ease some sort of the confusion if only part of inner transaction array returns. Adding a new filter param to return aggregate / embedded complete set as it used to be (as @fboucquez suggested) would also be a good choice

**Comment by Vektrat** [comment]
the point of this design is not that the user has to hope for the aggregates to be in the same page, or to keep querying for pages to fill it, the point is that if the user wants to fetch aggregate information, he needs to call the detail endpoint, and he will have all the information he needs, even to serialize it, because detail call does include the `transactions` array

i agree with that having a `numEmbeddedTransactions` attribute would help, i think i mentioned this to david, but unsure what the outcome of the discussion was, though either way it's not really up to us but up to mongodb (core)

**Comment by fboucquez** [comment]
Would it be possible to add the size attribute to the transactions? Users can calculate effective fees and recommended fees without asking for all the data.

**Comment by Vektrat** [comment]
adding the attribute is something that needs to be discussed, i remember asking this to david weeks ago and he was not for it - and it would be troublesome to add it from our side because other than more queries, it probably involves again completing queries manually, and it modifies the original database entity which we always prefer to avoid doing

i don't remember if the case of an empty aggregate is valid, probably not and may have been one of the reasons why david went with those decisions

my take on this problem is still the same: the list endpoint is a generic way of getting an overview of the transactions, if the client wants more information, he has to call the detail endpoint on the specific transactions (whether it be to get info, to cosign, to serialize or to just display it), easy as that, and we don't mind if he has to make extra queries because this is natural, it's a standard design, and it does not have any performance implications (in fact if there were, they would be less than before since queries at the REST side are WAY lighter than before, and at the client side are almost the same and still pretty much ordinary for a REST client)

to recap: the clients list transactions to get an overall list, if they spot an aggregate and need it to be complete, they make another request to fill in the details be it empty (if possible), or full

**Comment by dgarcia360** [comment]
@fboucquez 

> The calculated size would be incorrect
> The effective_fee and recommended fees cannot be calculated as the size is incorrect

Good catch. We thought about adding ``numTransactions`` and ``numCosignatories`` to the response, but assumed REST would have to query all related transactions and this is something we were trying to avoid.

 I like the idea of showing the transaction size under ``meta`` for every transaction, not just for the aggregate transaction. This value could be used to compare the size being calculated by the SDKs. Does rest know about the transaction ``size``? @Vektrat @martinayora If not, we could ask the core devs if it's possible to add it in the next release.

**Comment by Vektrat** [comment]
we can ask them, and obviously it will be returned then

but it was already discussed and accepted that returned transactions in the list are not to be treated as complete entities, thus the clients need to get a single transaction (aggregate) in order to get all the info, so things should be fine as they are - we also determined that operating around fees when listing transactions was not to happen / is not happening anyway, so this use case is far fetched

**Comment by Vektrat** [comment]
closing this because it isn't an issue, additionally, core team added `size` as a transaction field
